instrumented-tests/http-server-mock/Sources/HTTPServerMock/ServerSession.swift

```diff
@@ -35,50 +21,38 @@ public class ServerSession {
         self.recordingURL = server.baseURL.appendingPathComponent(sessionIdentifier)
     }
 
-    /// Fetches details of all `POST` requests recorded by the server during this session.
-    public func getRecordedPOSTRequests() throws -> [POSTRequestDetails] {
+    /// Returns all requests recorded by the server in this session.
+    public func getRecordedRequests() throws -> [Request] {
         return try server
-            .getRecordedPOSTRequestsInfo() // get all recorded requests info
-            .filter { requestInfo in requestInfo.path.contains(sessionIdentifier) } // narrow it to this session
-            .map { requestInfo in
-                return POSTRequestDetails(
-                    path: requestInfo.path,
-                    httpHeaders: try server.getRecordedRequestHeaders(requestInfo),
-                    httpBody: try server.getRecordedRequestBody(requestInfo)
-                )
-            }
+            .getRecordedRequests() // get all recorded requests info
+            .filter { request in request.path.contains(sessionIdentifier) } // narrow it to this session
     }
 
-    /// Actively fetches 'POST` requests recorded by the server until a desired count is found, or timeouts returning current recorded requests
-    public func pullRecordedPOSTRequests(count: Int, timeout: TimeInterval) throws -> [POSTRequestDetails] {
-        var currentRequests = [ServerMock.RequestInfo]()
-
+    /// Actively fetches requests recorded by the server in this session until given `condition` evaluated to `true`.
+    /// Throws an exception if given `timeout` is exceeded.
+    public func pullRecordedRequests(timeout: TimeInterval, until condition: ([Request]) throws -> Bool) throws -> [Request] {
         let timeoutTimer = DispatchSource.makeTimerSource(queue: DispatchQueue.global())
         timeoutTimer.setEventHandler { timeoutTimer.cancel() }
         timeoutTimer.schedule(deadline: .now() + timeout, leeway: .nanoseconds(0))
         if #available(iOS 10.0, *) {
             timeoutTimer.activate()
         }
 
+        var pulledRequests: [Request] = []
+        var conditionMet = false
         repeat {
-            currentRequests = try server
-                .getRecordedPOSTRequestsInfo()
-                .filter { requestInfo in requestInfo.path.contains(sessionIdentifier) }
+            pulledRequests = try getRecordedRequests()
+            conditionMet = try condition(pulledRequests)
             Thread.sleep(forTimeInterval: 0.2)
-        } while !timeoutTimer.isCancelled && currentRequests.count < count
+        } while !timeoutTimer.isCancelled && !conditionMet
```

Comment:
this can simplified: `while timeoutTimer.isCancelled || conditionMet`

Reply:
Actually, it becomes `!(timeoutTimer.isCancelled || conditionMet)` but it's still better, done 👍.